elastic/apm-server
# processor/otel: adjust timestamps for mobile

## Motivation/summary

Add support for a custom resource attribute, `telemetry.sdk.elastic_export_timestamp`, representing the time at which the agent sent the event payload to the server.

The server uses this to adjust the timestamps of events, to cater for end-user devices having incorrect system clock settings. When the server receives an OTLP payload, it records its current system time as `receive_timestamp`. Then it calculates `time_delta = receive_timestamp - export_timestamp`, and adds this to the timestamp of each event in the payload.

For example, let's say the client's clock is 5 minutes behind. It starts a span at 12:00, ends it at 12:05, and exports it to the server at 12:10; the server receives it at 13:10. The server records:
 - `receive_timestamp = 13:10`
 - `export_timestamp = 12:10`
 - `time_delta = 13:10-12:10 = 1h`

This is then added to the timestamp of the span, adjusting it to 13:00.

One important thing to note: this change means the network latency will always be included in the timestamp. If the client's clock is correct and network latency is significant, then we could increase error. We may want to later introduce a time delta threshold for the adjustment, i.e. don't apply a delta if it's less than Xms.

## Checklist

~- [ ] Documentation has been updated~

## How to test these changes

1. Once https://github.com/elastic/apm-agent-ios/pull/36 is merged, create some distributed traces with the iOS agent and opbeans-*
2. Check that they show up correctly on the waterfall.

## Related issues

Closes https://github.com/elastic/apm-server/issues/5173
